cpp/src/arrow/array.cc

```diff
@@ -304,6 +305,55 @@ Status ListArrayFromArrays(const Array& offsets, const Array& values, MemoryPool
   return Status::OK();
 }
 
+static std::shared_ptr<Array> SliceArrayWithOffsets(const Array& array, int64_t begin,
+                                                    int64_t end) {
+  return array.Slice(begin, end - begin);
+}
+
+template <typename ListArrayT>
+Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+                                                MemoryPool* memory_pool) {
+  const int64_t list_array_length = list_array.length();
+  std::shared_ptr<arrow::Array> value_array = list_array.values();
+
+  // Shortcut: if a ListArray does not contain nulls, then simply slice its
+  // value array with the first and the last offsets. We don't use
+  // Array::null_count() because it's potentially O(N) (thus not faster than not
+  // taking the shortcut).
```

Comment:
This is suboptimal. `null_count()` is cached, and even the first computation is much faster than walking the null bits one by one (because it uses CPU popcount instructions over entire words).

Reply:
Good to know. Thanks!